src/qibolab/instruments/rfsoc.py

```diff
@@ -182,14 +279,14 @@ def _open_connection(host: str, port: int, server_commands: dict):
                 if not tmp:
                     break
                 received.extend(tmp)
-        results = pickle.loads(received)
-        if isinstance(results, Exception):
-            raise results
+        results = json.loads(received)
+        if isinstance(results, str):
+            raise RuntimeError(results)
```

Comment:
This check is not any longer required: if it is not an appropriate object, it will fail when you use it.

E.g. at the moment it could be `list`, or an `int`, that would be equally wrong, but still passing this test.

In Python, at least, it is customary to adopt the duck typing approach: for as long as it works, don't care too much about checking what it is, it's working :)
(in this case, you need a subscribable object, with keys `"i"` and `"q"` - if it's a `dict` or not you don't care).

Reply:
The idea was that, whatever error is raised by the server, it gets sent to the client and raised there. With pickle this line was `if isinstance(results, Exception)`.

Reply:
I remember how it was with pickle, but them, if you want to keep raising the errors this way, be more specific:
- define your own `QibosoqError` (perhaps subclassing `RuntimeErorr`), and
- catch a marker in the string, e.g. `if isinstance(...) and "myerror" in results`...